server/resp/src/main/java/org/infinispan/server/resp/commands/ArgumentUtils.java

```diff
@@ -90,25 +90,38 @@ public static long toLong(byte[] argument) {
       if (argument[0] < '0') {
          if ((argument[0] != '-' && argument[0] != '+') || argument.length == 1)
             throw new NumberFormatException("Invalid character: " + argument[0]);
-
-         negative = true;
+         if (argument[0] == '-') {
+            negative = true;
+         }
          i = 1;
       }
 
-      long result;
+      long result=0;
       byte b = argument[i++];
       if (b < '0' || b > '9')
          throw new NumberFormatException("Invalid character: " + b);
+      // Summing negative number (48-b) so we can handle -9223372036854775808
+      result = (result << 3) + (result << 1) + (48 - b);
 
-      result = (b - 48);
       for (; i < argument.length; i++) {
          b = argument[i];
          if (b < '0' || b > '9')
             throw new NumberFormatException("Invalid character: " + b);
+         result = (result << 3) + (result << 1) + (48 - b);
+         // Check if overflow happened, result*10 turns > 0
+         if (result >= -9) {
+            throw new NumberFormatException("Value out of range: " + ArgumentUtils.toNumberString(argument));
+         }
+      }
 
-         result = (result << 3) + (result << 1) + (b - 48);
+      if (negative) {
```

Comment:
Should it be `!negative`?

Reply:
no, `result` contain the opposite of the value, because the sum is over the negative of the digit `48-b`, this to accept  '-9223372036854775808'